Marlin/Regulator.cpp

```diff
@@ -33,4 +69,98 @@ void setOutputPressure(float desired_pressure) {
     DAC_write(MCP4725_I2C_ADDRESS, digital_val);
 }
 
+
+/**
+ * Updates the protection function, called regularly in the main loop.
+ */
+void Regulator__Update() {
+  pressure_protection(current_target_pressure, pressureRegulator());
+}
+
+//===========================================================================
+//============================ Private Functions ============================
+//===========================================================================
+
+/** 
+ * Error handler when marlin detects a missing pressure regulator
+ */
+static void _regulator_sled_removed_error() {
+  if (sled_present == true) {
+    //quickStop();
+    //disable_all_heaters();
+    //disable_all_steppers();
+    SERIAL_PROTOCOLLNPGM(  
+      "Sled not detected. Canceling print, check pneumatics sled.");
+    SERIAL_PROTOCOLLNPGM("// action:cancel");
+    sled_present = false;
+  }
+}
+
+/** 
+ * Error handler when marlin detects a regulator leak 
+ */
+static void _regulator_leak_error() {
+  quickStop();
+  disable_all_heaters();
+  disable_all_steppers();
+  SERIAL_PROTOCOLLNPGM(  
+      "Pressure Leak Error. Canceling print, check the pressure hose");
+  SERIAL_PROTOCOLLNPGM("// action:cancel");
+  Regulator__SetOutputPressure(0.0);
+  sled_present = true;
+}
+
+/** 
+ * Error handler when marlin detects a regulator pressure that's too high
+ */
+static void _regulator_runaway_error() {
+  quickStop();
+  disable_all_heaters();
+  disable_all_steppers();
+  SERIAL_PROTOCOLLNPGM("Runaway Pressure Error. Canceling print.");
+  SERIAL_PROTOCOLLNPGM("// action:cancel");
+  Regulator__SetOutputPressure(0.0);
+  sled_present = true;
+}
+
+/** 
+ * Allows us to determine if the pressure is too high or low, indicating
+ * a serious issue. Checks the target temperature vs. the desired, and 
+ * throws an error if outside the bounds for too long.
+ * @target_pressure  The pressure you're attempting to reach
+ * @pressure         The actual current pressure 
+ */
+static void pressure_protection(float target_pressure, float pressure) {
+  static float prev_target_pressure = 0;
+  static millis_t regulatorTimer = 0;
+
+  if (regulatorTimer == 0) regulatorTimer = millis();
+  bool regulatorUnder =
+      (pressure <= (target_pressure - REGULATOR_PROTECTION_BAND));
+  bool regulatorOver =
+      (pressure >= (target_pressure + REGULATOR_PROTECTION_BAND));
+
+  if (regulatorOver || regulatorUnder) {
+    if (millis() > (regulatorTimer + REGULATOR_PROTECTION_TIME_S * 1000UL)) {
```

Comment:
This is probably optimized out anyways, but should you just make REGULATOR_PROTECTION_TIME_MS